### PR TITLE
Mk transifex pull

### PR DESCRIPTION
**Pull request by mkangia**
Second phase for transifex integration after https://github.com/dimagi/commcare-hq/pull/20492
This PR:
1. provides the view changes to pull translated files from transifex
2. checks if all has been translated before pulling files. Provided an option for one to skip that test if needed for now
3. Generates an excel file just like one would download for app translations and emails it to the user

Probably better to review all at once.

### CODE REVIEW

File: custom/icds/translations/integrations/transifex.py [hunk 1]
```diff
@@ -83,3 +93,64 @@ def _cleanup(self):
         for resource_name, filepath in self.po_file_generator.generated_files:
             if os.path.exists(filepath):
                 os.remove(filepath)
+
+    def _get_resource_slugs_for_version(self, version):
```
Comment (esoergel) [inline custom/icds/translations/integrations/transifex.py]: It looks like `version` is always just `self.version`.  Would it be possible to make a memoized property that's just `resource_slugs`?  Then you wouldn't need to pass it around as much.

File: custom/icds/translations/integrations/transifex.py [hunk 2]
```diff
@@ -83,3 +93,64 @@ def _cleanup(self):
...
+                ))
+
+    def get_translations(self, resource_slugs=None):
+        """
+        pull translations from transifex
+        :param resource_slugs: optional argument. All resource slugs corresponding to version are used otherwise.
+        :return: list of POEntry objects
+        """
+        if resource_slugs:
+            self._ensure_resource_slugs_for_version(resource_slugs, self.version)
```
Comment (esoergel) [inline custom/icds/translations/integrations/transifex.py]: I think this check should go in the code that fetches the `resource_slugs` - one place that's responsible for returning valid slugs, so everywhere else knows what they're working with.

File: corehq/apps/app_manager/app_translations/parser.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import re
+from datetime import datetime
+from openpyxl import Workbook
+from tempfile import NamedTemporaryFile
+
+from corehq.apps.app_manager.app_translations.const import MODULES_AND_FORMS_SHEET_NAME
+from corehq.apps.dump_reload.const import DATETIME_FORMAT
+
+
+class TranslationsParser:
```
Comment (esoergel) [inline corehq/apps/app_manager/app_translations/parser.py]: I think this should be
```python
class TranslationsParser(object):
```

File: custom/icds/translations/integrations/transifex.py [hunk 4]
```diff
@@ -26,14 +28,21 @@ def __init__(self, domain, app_id, source_lang, project_slug, version=None, lang
...
+        self.version = version
+        self.key_lang = "en"  # the lang in which the string keys are, should be english
+        self.lang_prefix = lang_prefix
         self.project_slug = project_slug
         self.is_source_file = is_source_file
         self.source_lang = source_lang
-        self.po_file_generator = POFileGenerator(domain, app_id, version, key_lang, source_lang, lang_prefix,
+        self.po_file_generator = POFileGenerator(domain, app_id, version, self.key_lang, source_lang, lang_prefix,
                                                  exclude_if_default)
+        self.parser = TranslationsParser(self)
```
Comment (esoergel) [inline custom/icds/translations/integrations/transifex.py]: This dependency is a little weird - `Transifex` has an attribute called `parser`, which itself has `transifex` as an attribute.
It looks like `transifex.parser` is only called once, for `generate_excel_file`.  An alternative approach would be to make a `generate_excel_file` method on the `Transifex` class like this:
```python
def generate_excel_file(self):
    parser = TranslationsParser(self)
    return parser.generate_excel_file()
```
That way you could change the parser around without needing to worry about the calling code, since it doesn't access it directly.  Further, my preference would be to pass only the dependencies that `TranslationsParser` needs, rather than giving it this whole class.  That would avoid such tight coupling between the two classes.

File: custom/icds/translations/integrations/transifex.py [hunk 5]
```diff
@@ -83,3 +93,64 @@ def _cleanup(self):
...
+            if not resource_slug.endswith("v%s" % version):
+                raise Exception("Resource name '{}' is expected to contain version".format(
+                    resource_slug
+                ))
+
+    def get_translations(self, resource_slugs=None):
+        """
+        pull translations from transifex
+        :param resource_slugs: optional argument. All resource slugs corresponding to version are used otherwise.
+        :return: list of POEntry objects
```
Comment (esoergel) [inline custom/icds/translations/integrations/transifex.py]: nit: it returns a dict, not a list

File: corehq/apps/app_manager/app_translations/parser.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        return ("TransifexTranslations {}-{}:v{} {}.xlsx".format(
+            self.transifex.key_lang, self.transifex.source_lang,
+            version, datetime.utcnow().strftime(DATETIME_FORMAT))
+        )
+
+    def _generate_sheets(self, wb, resource_slugs):
+        version = self.transifex.version
+        all_translations = self.transifex.get_translations(version, resource_slugs)
+        for resource_name in all_translations:
+            po_entries = all_translations[resource_name]
```
Comment (esoergel) [inline corehq/apps/app_manager/app_translations/parser.py]: You could also do:
```python
for resource_name, po_entries in all_translations.items():
```
or just make `all_translations` a list of tuples if you don't need the dictionary functionality anyways.

**Review by esoergel — COMMENTED** [review]
Looks like you're passing around tempfiles in a couple places here.  As we recently ran into with exports, you'll need to manually clean these up, including instances where there's an exception thrown somewhere in the middle.  The default way to do this is using `NamedTemporaryFile` in a `with` statement (but without `delete=False`).  This with statement should then span the whole lifecycle of that file.  Once you're done writing, you can then `seek` to 0 on the file and read from the same object.

I also wrote an alternative utility called [`TransientTempfile`](https://github.com/dimagi/commcare-hq/blob/master/corehq/util/files.py#L47-L47) which uses a `with` statement only for the lifecycle of the file, and you handle reads and writes with normal `with open` statements.

There are also a few optional parameters in here that I don't think are necessary.  If you make those mandatory, then you can eliminate a bunch of branching logic that deals with whether or not that was provided.  Essentially, it's a way to make your code stricter and a bit simpler.

File: corehq/apps/app_manager/app_translations/parser.py (hunk 3 above)
Comment (mkangia) [reply corehq/apps/app_manager/app_translations/parser.py]: hmm, i have started doing this recently. Is there an advantage by using object here?

File: corehq/apps/app_manager/app_translations/parser.py (hunk 6 above)
Comment (mkangia) [reply corehq/apps/app_manager/app_translations/parser.py]: I remember using it and it was causing an issue. let me try again.

list of tuples? How would that be?
need the dict where resource_name is mapped to a list of translations/PoEntries

File: corehq/apps/app_manager/app_translations/parser.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            sheet_name = resource_name.split("_v%s" % version)[0]
+            self.translations[sheet_name] = []
+            ws = wb.create_sheet(title=sheet_name)
+            self._parse_sheet(ws, po_entries)
+
+    def generate_excel_file(self, resource_slugs=None):
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/app_manager/app_translations/parser.py]: Ya, its more of a proactive addition because it would probably come in handy to pull specific resource slugs.  I used it during testing so i left it there for now. Happy to remove if you think its not necessary/YAGNI.

File: corehq/apps/app_manager/app_translations/parser.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        version = self.transifex.version
+        wb = Workbook(write_only=True)
+        self._generate_sheets(wb, resource_slugs)
+
+        with NamedTemporaryFile(delete=False) as tempfile:
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/app_manager/app_translations/parser.py]: Yup, done later.

File: custom/icds/translations/integrations/transifex.py (hunk 4 above)
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: that makes sense.. that circular reference was not a good plan. Updating.

File: custom/icds/translations/integrations/client.py [hunk 9]
```diff
@@ -66,3 +69,31 @@ def project_details(self):
...
+        return self.resource_details(resource_slug, lang)['completed'] == "100%"
+
+    def get_translation(self, resource_slug, lang):
+        url = "https://www.transifex.com/api/2/project/{}/resource/{}/translation/{}/?file".format(
+            self.project, resource_slug, lang
+        )
+        response = requests.get(url, auth=self._auth, stream=True)
+        if response.status_code != 200:
+            return response
+        temp_file = tempfile.NamedTemporaryFile()
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/client.py]: Just to confirm
```
temp_file = tempfile.NamedTemporaryFile()
temp_file.write(response.content)
temp_file.seek(0)
polib_obj = polib.pofile(temp_file.name)
temp_file.close()
```
This looks good?

File: custom/icds/translations/integrations/transifex.py (hunk 1 above)
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: Ya, removed passing version argument.
> memoized property that's just resource_slugs

sorry, could not get this.

File: custom/icds/translations/integrations/transifex.py (hunk 2 above)
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: It is called with resource_slugs as `None`. So if there are no resource slugs it tries to find them first and if its still missing, then it raises an `Exception`.

File: custom/icds/translations/integrations/transifex.py [hunk 10]
```diff
@@ -83,3 +93,64 @@ def _cleanup(self):
...
+        """
+        if resource_slugs:
+            self._ensure_resource_slugs_for_version(resource_slugs, self.version)
+        client = self.client()
+        if not resource_slugs:
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: It is called with resource_slugs as `None`. So if there are no resource slugs it tries to find them first and if its still missing, then it raises an Exception.

File: corehq/apps/app_manager/app_translations/parser.py (hunk 6 above)
Comment (mkangia) [reply corehq/apps/app_manager/app_translations/parser.py]: updated to use the for loop with items.

File: custom/icds/translations/integrations/transifex.py (hunk 2 above)
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: `resource_slugs` can be passed to this function as an argument. Currently there is no place where its fetched and then passed to this method.

**Comment by mkangia** [comment]
@esoergel 
Made few changes for the reviews and replied to rest.
Looks like there is some possibility of improving the structure for this. Will try to make some improvements.

File: custom/icds/translations/integrations/client.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (esoergel) [reply custom/icds/translations/integrations/client.py]: Actually I don't think you'll need the `seek` line.  Since you're not passing the open file reference to `polib`, it must open the file again.  Otherwise that looks correct.

You'll need to make sure `tempfile.close()` doesn't get closed until `polib_obj` is no longer needed, since it'll delete the file.  Similarly, it's safer to put all operations on this temfile inside a `with` statement so the file gets cleaned up even on failure.  That can look like this:
```python
with tempfile.NamedTemporaryFile() as tempfile:
    temp_file.write(response.content)
    polib_obj = polib.pofile(temp_file.name)
```
The downside of this is that `polib_obj` maintains a separate handler to the same file, so you'll have the file open twice.  I believe having a file open multiple times works on most systems (just not Windows, I think?).  It could get confusing if both tried writing to it in sequence.  That `TransientTempfile` util I mentioned in my review comment lets you close the file without yet deleting it.  This use-case is probably simple enough that it won't run into issues either way, though.

File: custom/icds/translations/integrations/transifex.py (hunk 1 above)
Comment (esoergel) [reply custom/icds/translations/integrations/transifex.py]: haha, it took me a minute to decipher my own words 3 days later here.  I meant that you could replace
```python
def _get_resource_slugs_for_version(self, version):
```
with
```python
@property
@memoized
def resource_slugs(self):
```

File: corehq/apps/app_manager/app_translations/parser.py (hunk 6 above)
Comment (esoergel) [reply corehq/apps/app_manager/app_translations/parser.py]: > list of tuples? How would that be?

I mean that these two are equivalent:

```python
po_entries = {}
for resource_slug in resource_slugs:
    po_entries[resource_slug] = self.client.get_translation(resource_slug, self.source_lang, self.lock_translations)
return po_entries

for resource_name, po_entries in get_translations().items():
```

```python
po_entries = [] 
for resource_slug in resource_slugs:
    po_entries.append((resource_slug, self.client.get_translation(resource_slug, self.source_lang, self.lock_translations)))
return po_entries

for resource_name, po_entries in get_translations():
```
Using a dictionary lets you jump to specific entries in the set, or more easily modify them after the fact, but in this case, it iterates over the whole set using `items()`, which converts the dictionary into a list of tuples, so you could also just build a list of tuples directly.  They're really pretty much the same though - I don't think it matters.

File: custom/icds/translations/integrations/client.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/client.py]: > Actually I don't think you'll need the seek line.

I tried this and just using `write` didnt write to file, had to do `seek` to make it actually write the content.

> You'll need to make sure tempfile.close() doesn't get closed until polib_obj is no longer needed, since it'll delete the file

Ya i think we don't need that file after that. `polib.pofile(temp_file.name)` parses the file and converts it to `POEntries`. i have updated this now and there is no `po_obj`.

[the remark this answers is not in the document]
Comment (esoergel) [reply custom/icds/translations/integrations/client.py]: > I tried this and just using write didnt write to file, had to do seek to make it actually write the content.

`seek(0)` moves the pointer back to the beginning of the file.  `write` does write to the file, but you can't read from the same file object again without resetting the pointer.  `polib.pofile` only gets the file name, so it must be opening the file a second time, which means it *should* be able to read without seeking.  Here's an example:
```python
In [1]: import tempfile

In [2]: with tempfile.NamedTemporaryFile() as f:
   ...:     f.write('hello')
   ...:     print "read 1", f.read()
   ...:     f.seek(0)
   ...:     print "read 2", f.read()
   ...:     print "read 3", f.read()
   ...:     with open(f.name) as f2:
   ...:         print "read 4", f2.read()
   ...:         
read 1 
read 2 hello
read 3 
read 4 hello
```

> Ya i think we don't need that file after that. polib.pofile(temp_file.name) parses the file and converts it to POEntries. i have updated this now and there is no po_obj.

Cool yeah, as long as that doesn't need the tempfile after initialization this should be fine.  I wasn't sure whether it streamed the file lazily or loaded it all into memory up front, for example.

File: custom/icds_reports/views.py [hunk 11]
```diff
@@ -1624,13 +1624,13 @@ def page_context(self):
...
-        source_language_code = form_data.get('target_lang') or form_data.get('source_lang')
-        return Transifex(domain, form_data['app_id'], source_language_code, transifex_project_slug,
-                         form_data['version'])
+    def transifex(self, domain, form_data):
+        if not hasattr(self, '_transifex'):
+            transifex_project_slug = form_data.get('transifex_project_slug')
+            source_language_code = form_data.get('target_lang') or form_data.get('source_lang')
+            self._transifex = Transifex(domain, form_data['app_id'], source_language_code, transifex_project_slug,
+                                        form_data['version'])
+        return self._transifex
```
Comment (esoergel) [inline custom/icds_reports/views.py]: Looks like this is only accessed directly once, so you could just rename it `get_transifex` and not bother with memoizing the result.

**Review by esoergel — APPROVED** [review]

File: custom/icds/translations/integrations/transifex.py [hunk 12]
```diff
@@ -83,3 +88,68 @@ def _cleanup(self):
...
+        # if still no resources raise error
+        if not resource_slugs:
+            raise Exception("No resources found for this version")
+        po_entries = {}
+        for resource_slug in resource_slugs:
+            po_entries[resource_slug] = self.client.get_translation(resource_slug, self.source_lang,
+                                                                    self.lock_translations)
+        return po_entries
+
+    def resources_pending_translations(self, break_if_true=False):
```
Comment (esoergel) [inline custom/icds/translations/integrations/transifex.py]: This method returns a single slug if `break_if_true`, and a list of slugs if `break_if_false`.  It looks like this is used in only one place, and it's being used as a boolean.  It feels to me like it's one method trying to act like several depending on how it's called.  Considering this is only used in a boolean context, and with `break_if_true=True`, I'd change the name and implementation to reflect that.  That's my preference, but I'll defer to you on the final call.

File: custom/icds/translations/integrations/client.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/translations/integrations/client.py]: @esoergel tried making this change and ran into some issue with encoding because of it.
Want to leave it as-is for now and look at this improvement later.

Also looks like with the current implementation,
it should be safe to leave as-is. When the `temp_file` reference is removed, the file will get removed as well, so there should be no orphan file references, temp files even if anything goes wrong. let me know if thats okay.

File: custom/icds/translations/integrations/transifex.py (hunk 12 above)
Comment (mkangia) [reply custom/icds/translations/integrations/transifex.py]: > it's being used as a boolean

the result is used to list the pending resources/resources as well, so `resources_pending_translations` sounds fine. Can make it always return a list, so in case `break_if_true` is `True` return a list with the single resource to make the return predictable.

File: custom/icds/translations/integrations/transifex.py [hunk 13]
```diff
@@ -89,42 +91,33 @@ def _cleanup(self):
...
-        return [r['name']
-                for r in self.client.list_resources().json()
-                if r['name'].endswith("v%s" % self.version)]
+    @property
+    @memoized
+    def resource_slugs(self):
+        if self._resource_slugs:
+            return self._resource_slugs
+        else:
+            return self.client.get_resource_slugs(self.version)
```
Comment (esoergel) [inline custom/icds/translations/integrations/transifex.py]: :+1:

**Review by esoergel — APPROVED** [review]
